### PR TITLE
fix(AzureVpcPeering): fixes deletion issues when Kyma SPN is unauthorized

### PR DESCRIPTION
**Pull request by vandjelk**
**Description**

Changes proposed in this pull request:

- fixes deletion issues when Kyma SPN is unauthorized in remote subscription
- adds peering loading fail-over
- relaxes reconciler requeuing on Unauthorized errors

### CODE REVIEW

File: pkg/kcp/provider/azure/vpcpeering/deleteVpcPeering.go [hunk 1]
```diff
@@ -4,34 +4,41 @@ import (
...
 
 func deleteVpcPeering(ctx context.Context, st composed.State) (error, context.Context) {
 	state := st.(*State)
 	logger := composed.LoggerFromCtx(ctx)
 	obj := state.ObjAsVpcPeering()
 
-	if len(obj.Status.Id) == 0 {
-		logger.Info("VpcPeering deleted before Azure peering is created")
+	if state.localPeering == nil {
+		logger.Info("Skipping local peering deletion")
```
Comment (dushanpantic) [inline pkg/kcp/provider/azure/vpcpeering/deleteVpcPeering.go:17]: If my domain understanding is correct, if we enter this block, the peering is already deleted. Under that assumption, this message is misleading.

File: pkg/kcp/provider/azure/vpcpeering/deleteVpcPeering.go [hunk 2]
```diff
@@ -4,34 +4,41 @@ import (
...
 		return nil, nil
 	}
 
-	resourceGroupName := state.Scope().Spec.Scope.Azure.VpcNetwork
+	resourceId, err := util.ParseResourceID(ptr.Deref(state.localPeering.ID, ""))
+
+	if err != nil {
+		logger.Error(err, "Failed parsing localPeering.ID while deleting local peering")
+		return nil, nil
+	}
```
Comment (dushanpantic) [inline pkg/kcp/provider/azure/vpcpeering/deleteVpcPeering.go:26]: Important: Local peering must be deleted. We must not continue the reconcile in a case of error.

File: pkg/kcp/provider/azure/vpcpeering/deleteVpcPeering.go [hunk 3]
```diff
@@ -4,34 +4,41 @@ import (
...
-		resourceGroupName,
-		state.Scope().Spec.Scope.Azure.VpcNetwork,
+		resourceId.ResourceGroup,
+		resourceId.ResourceName,
 		obj.GetLocalPeeringName(),
 	)
 
 	if err != nil {
-		return azuremeta.LogErrorAndReturn(err, "Error deleting vpc peering", ctx)
+		return azuremeta.LogErrorAndReturn(err, "Error deleting local vpc peering", ctx)
```
Comment (dushanpantic) [inline pkg/kcp/provider/azure/vpcpeering/deleteVpcPeering.go:38]: This azuremeta.LogErrorAndReturn is hiding behavior.
Consider not using it for return values (outside of this PR)

File: pkg/kcp/provider/azure/vpcpeering/peeringLocalCreate.go [hunk 4]
```diff
@@ -45,20 +45,25 @@ func peeringLocalCreate(ctx context.Context, st composed.State) (error, context.
 			state.ObjAsVpcPeering().Status.State = string(cloudcontrolv1beta1.ErrorState)
 		}
 
+		condition := metav1.Condition{
+			Type:    cloudcontrolv1beta1.ConditionTypeError,
+			Status:  metav1.ConditionTrue,
+			Reason:  cloudcontrolv1beta1.ReasonFailedCreatingVpcPeeringConnection,
+			Message: message,
+		}
```
Comment (dushanpantic) [inline pkg/kcp/provider/azure/vpcpeering/peeringLocalCreate.go:53]: Couple lines above this you are setting `.status.state` to warning or error.
Please reconsider if its necessary to set this condition to `ConditionTypeWarning` in a case when `.status.state` is set to Warning as well.

File: pkg/kcp/provider/azure/vpcpeering/peeringLocalLoad.go [hunk 5]
```diff
@@ -4,25 +4,45 @@ import (
...
+
+	if !ok && state.localNetworkId != nil {
+		resourceGroup = state.localNetworkId.ResourceGroup
+		networkName = state.localNetworkId.NetworkName()
+		ok = true
+	}
+
+	if !ok {
 		return nil, nil
 	}
```
Comment (dushanpantic) [inline pkg/kcp/provider/azure/vpcpeering/peeringLocalLoad.go:36]: Consider logging info when no local peering is loaded

File: pkg/kcp/provider/azure/vpcpeering/peeringRemoteLoad.go [hunk 6]
```diff
@@ -19,11 +20,34 @@ func peeringRemoteLoad(ctx context.Context, st composed.State) (error, context.C
...
+
+	if !ok && state.localNetworkId != nil {
+		resourceGroup = state.remoteNetworkId.ResourceGroup
+		networkName = state.remoteNetworkId.NetworkName()
+		ok = true
+	}
+
+	if !ok {
+		return nil, nil
+	}
```
Comment (dushanpantic) [inline pkg/kcp/provider/azure/vpcpeering/peeringRemoteLoad.go:44]: Consider logging info when remote peering is not loaded

File: pkg/kcp/provider/azure/vpcpeering/peeringRemoteDelete.go [hunk 7]
```diff
@@ -13,21 +15,27 @@ func peeringRemoteDelete(ctx context.Context, st composed.State) (error, context
 		return nil, nil
 	}
 
-	if len(state.ObjAsVpcPeering().Status.RemoteId) == 0 {
-		logger.Info("Remote VpcPeering deleted before Azure peering is created")
+	if state.remotePeering == nil {
+		logger.Info("Skipping remote peering deletion")
```
Comment (dushanpantic) [inline pkg/kcp/provider/azure/vpcpeering/peeringRemoteDelete.go:19]: Perhaps a better info message would be something along lines "remote peering not loaded, continuing".
Current message might make it seem like we know there is a peering, but we are choosing not to delete it.

File: internal/controller/cloud-control/vpcpeering_azure_test.go [hunk 8]
```diff
@@ -521,4 +521,143 @@ var _ = Describe("Feature: KCP VpcPeering", func() {
...
+						metav1.ConditionTrue,
+						cloudcontrolv1beta1.ReasonFailedCreatingVpcPeeringConnection,
+						"Not authorized to perform action."),
+				).
+				Should(Succeed())
+		})
+
+		// DELETE
+
+		By("When KCP VpcPeering is deleted", func() {
```
Comment (dushanpantic) [inline internal/controller/cloud-control/vpcpeering_azure_test.go:624]: every step above this should be `Given`

File: pkg/kcp/provider/azure/vpcpeering/deleteVpcPeering.go (hunk 1 above)
Comment (vandjelk) [reply pkg/kcp/provider/azure/vpcpeering/deleteVpcPeering.go:17]: In that case I will leave old message "VpcPeering deleted before Azure peering is created"

File: pkg/kcp/provider/azure/vpcpeering/deleteVpcPeering.go (hunk 2 above)
Comment (vandjelk) [reply pkg/kcp/provider/azure/vpcpeering/deleteVpcPeering.go:26]: I will log and return error

**Review by dushanpantic — APPROVED** [review]

**Pull request merged**